src/Nethermind/Nethermind.TxPool/TxPool.cs

```diff
@@ -338,6 +310,171 @@ private AddTxResult AddCore(Transaction tx, bool isPersistentBroadcast, bool isR
+            
+            if (bucketSnapshot.Length == 0)
+            {
+                return;
+            }
+            
+            Account? account = _stateProvider.GetAccount(senderAddress);
+            UInt256 balance = account?.Balance ?? UInt256.Zero;
+            long currentNonce = (long)(account?.Nonce ?? UInt256.Zero);
+
+            UpdateGasBottleneck(bucketSnapshot, currentNonce, balance);
+        }
+
+        private void UpdateGasBottleneck(IReadOnlyList<Transaction> bucketSnapshot, long currentNonce, UInt256 balance)
+        {
+            Transaction tx = bucketSnapshot[0];
+            UInt256 previousTxBottleneck = tx.IsEip1559 ? CalculatePayableGasPrice(tx, balance) : tx.GasPrice;
+
+            for (int i = 0; i < bucketSnapshot.Count; i++)
+            {
+                tx = bucketSnapshot[i];
+                UInt256 gasBottleneck = 0;
+
+                if (tx.Nonce == currentNonce + i)
+                {
+                    UInt256 effectiveGasPrice = tx.IsEip1559 ? UInt256.Min(tx.FeeCap, tx.GasPremium + CurrentBaseFee) : tx.GasPrice;
+                    gasBottleneck = UInt256.Min(effectiveGasPrice, previousTxBottleneck);
+                }
+                
+                _transactions.NotifyChange(tx.Hash, tx, t =>
+                {
+                    tx.GasBottleneck = gasBottleneck;
+                });
+                
+                previousTxBottleneck = gasBottleneck;
+            }
+        }
+
+        private UInt256 CalculatePayableGasPrice(Transaction tx, UInt256 balance)
+        {
+            UInt256 payableGasPrice;
+            
+            if (balance > tx.Value)
+            {
+                UInt256 effectiveGasPrice = UInt256.Min(tx.FeeCap, tx.GasPremium + CurrentBaseFee);
+                UInt256 balanceAvailableForFeePayment = balance - tx.Value;
+                balanceAvailableForFeePayment.Divide((UInt256)tx.GasLimit, out UInt256 maxPayablePricePerGasUnit);
+                payableGasPrice = UInt256.Min(effectiveGasPrice, maxPayablePricePerGasUnit);
+            }
+            else
+            {
+                payableGasPrice = 0;
+            }
+
+            return payableGasPrice;
+        }
+
+        public void RemoveOrUpdateBucket(Address senderAddress)
+        {
+            var bucketSnapshot = _transactions.GetBucketSnapshot(senderAddress);
+
+            if (bucketSnapshot.Length == 0)
+            {
+                return;
+            }
+            
+            Account? account = _stateProvider.GetAccount(senderAddress);
+            UInt256 balance = account?.Balance ?? UInt256.Zero;
+            long currentNonce = (long)(account?.Nonce ?? UInt256.Zero);
+            Transaction tx = bucketSnapshot[0];
+
+            bool insufficientBalance = false;
+
+            if (balance < tx.Value)
+            {
+                insufficientBalance = true;
+            }
+            else if (!tx.IsEip1559)
```

Comment:
What about tx.IsEip1559?

Reply:
If balance < tx.value, there is insufficient balance no matter of tx type. If tx is legacy, balance need to be at least equal tx.value + tx.gasprice, but 1559 type needs only more than tx.value - it can be executed even if balance < tx.value + tx.feecap. Payable gasprice (lower value from effective gasprice and (balance - tx.value / tx.gaslimit)) is a gasbottleneck of this tx and is sorted by this value. No need to remove it at this point - has a chance of executing.